### DISCUSSION TITLE
Disable auto focus on dialog open

### DISCUSSION THREAD

**Discussion by kylerphillips** [shadcn-ui/ui#2307] [Q&A]
I'm using the default [Dialog](https://ui.shadcn.com/docs/components/dialog) component with input fields from the shadcn docs. 

I wanted to know if it's possible to disable the auto focus on inputs when the dialog opens?

**Comment by riculum** [comment]
The same question arises for the Sheet component 👍

**Comment by SoyDiego** [comment]
Maybe I'm wrong but I think is necessary for Accessibility.

**Reply by riculum** [reply]
Maybe but then it should be possible to change the first focus element

**Comment by PiotrWarzachowski** [comment]
I wanted to contribute a bit by making this open source project, where I used shadcn , you can contribute and customize some objects if you want, here is my repo : https://github.com/PiotrWarzachowski/DocMind Feel free to check it out, its public on vercel. You can leave a star ⭐ if you want 💓. Have a good year

**Comment by HereOrCode** [comment]
![](https://github.com/shadcn-ui/ui/assets/83153991/74043633-39e2-4371-be5b-e9243fd7d44f)

**Reply by JoyShaheb** [reply]
thank you @code4you2021  for the solution, even worked on the buttons on the `DialogFooter`

**Reply by villszer** [reply]
For easy copy pasta: `onOpenAutoFocus={(e) => e.preventDefault()}`

**Reply by kapxapot** [reply]
I tried this and it works, but when I tab through elements, the tab is not trapped inside of a dialog at first and goes through background elements and only then comes to the dialog and is trapped there. Is there a fix for that?

**Reply by soham2k06** [reply]
> I tried this and it works, but when I tab through elements, the tab is not trapped inside of a dialog at first and goes through background elements and only then comes to the dialog and is trapped there. Is there a fix for that?

Same problem but with form inputs, it focuses to the dialog in between

**Reply by LikeDreamwalker** [reply]
Well it is not working in my scenario, I can see the logs in onOpenAutoFocus event, but seems like it is not stopped, or some other part triggers the focus.

**Reply by S4hil-Yadav** [reply]
i wanna marry u bro

**Reply by lukebrandonfarrell** [reply]
Hey @S4hil-Yadav  he's mine

**Reply by S4hil-Yadav** [reply]
> Hey @S4hil-Yadav he's mine

You're gonna hav to fight me to death for that

**Reply by surajondev** [reply]
```js
onOpenAutoFocus={(e) => e.preventDefault()
 ```
 
 Easy to copy paste.

**Reply by truethari** [reply]
> ```js
> onOpenAutoFocus={(e) => e.preventDefault()
> ```

```js
onOpenAutoFocus={(e) => e.preventDefault()}
```

you missed a bracket

**Reply by rs-duongnt1** [reply]
Tks pro!!!!!!!!!

**Comment by Rezahy** [comment]
just add "tabIndex={-1}" on the input element.

**Comment by azizdevuz** [comment]
Vue.

> <DialogContent  @openAutoFocus="(e:Event) => e.preventDefault()"></DialogContent>

**Reply by Gwizdo51** [reply]
Even better :
```vue
<DialogContent @openAutoFocus.prevent>
```

**Reply by EriksonEnterprises** [reply]
> Even better :
> 
> ```vue
> <DialogContent @openAutoFocus.prevent>
> ```

This should be the accepted answer, thank you!

Also to note, I noticed when I don't have any fields in the dialog, sometimes it'll make some border around the dialog. I had to add this to the `:class="cn(`
'focus:outline-none',

**Comment by ystrdy** [comment]
Here's my solution:

Define the following `usePreventAutoFocus` hook:

```typescript
// @/hooks/use-prevent-auto-focus.ts
import { useRef, useCallback } from 'react';

function usePreventAutoFocus<E extends HTMLElement = HTMLDivElement>() {
  const ref = useRef<E>(null);

  const onOpenAutoFocus = useCallback((event: Event) => {
    event.preventDefault();
    ref.current?.focus({ preventScroll: true });
  }, []);

  return { ref, onOpenAutoFocus, tabIndex: -1 as const };
}

export default usePreventAutoFocus;
```

And here's how to use it in a dialog:

```tsx
import usePreventAutoFocus from '@/hooks/use-prevent-auto-focus'

const prevent = usePreventAutoFocus();

return (
  <Dialog open={open} onOpenChange={handleOpenChange}>
    <DialogContent {...prevent} className="outline-none">
      <DialogHeader>
        <DialogTitle>Share Folder</DialogTitle>
        <DialogDescription>
          Select the folder to share and the members to share it with.
        </DialogDescription>
      </DialogHeader>

      <div>
        <Label className="text-base font-semibold block mb-2">Select Folder</Label>
        <InputSearch
          placeholder="Search folders..."
          value={folderSearchTerm}
          onChange={(e) => setFolderSearchTerm(e.target.value)}
          className="my-2 h-9"
          disabled={sharing}
        />
      </div>
    </DialogContent>
  </Dialog>
);
```

**Reply by auzanxp** [reply]
thank you for solution bro, it's work

**Reply by kevincrossgrove** [reply]
Thank you, works perfectly <3

**Comment by aBenve** [comment]
I founded a solution for me at least, just add `tabIndex={undefined}` to the `<DialogContent />` (or the inner `<DialogPrimitive.Content />`) and you are good to go!

**Reply by mattanimation** [reply]
Thanks! This fixed it for me where I had a form inside a Content element it wouldn't allow me to tab to other inputs. none of the autofocus, onOpen or onClose worked.

**Comment by Prajwal-alphanimble** [comment]
you might wanna check out this discussion where I implemented a new resizable and dragable dashboard layout... this can be very helpfull to many developers in making analytic, sharable dashboard, the users can alter the dashboard based on their liking...

The disscussion link: https://github.com/shadcn-ui/ui/discussions/10466
